### ISSUE TITLE
1.16.0 raises Errno::ENOENT during `rails new`

### ISSUE THREAD

**Issue by jaredbeck** [bundler/bundler#6148]
# Error Report

```
Errno::ENOENT: No such file or directory @ rb_sysopen - /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/templates/Executable
...
Unfortunately, an unexpected error occurred, and Bundler cannot continue.
```

- **What did you do?**

  I ran the command `/Users/jared/.rbenv/versions/2.4.2/bin/bundle binstubs webpacker`

(Actually, I ran `rails new --skip-turbolinks --webpack=react --skip-spring --skip-action-cable --skip-coffee --skip-test --no-rc learn_webpacker`)

- **What did you expect to happen?**

  I expected Bundler to install rails

- **What happened instead?**

bundler crashed and asked me to file this bug report

- **Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**

After downgrading to 1.15.4 I can no longer reproduce the issue.

- **Have you read our issues document, https://github.com/bundler/bundler/blob/master/doc/contributing/ISSUES.md?**

Yes

  ...

## Backtrace

```
Errno::ENOENT: No such file or directory @ rb_sysopen - /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/templates/Executable
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/installer.rb:127:in `read'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/installer.rb:127:in `generate_bundler_executable_stubs'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/cli/binstubs.rb:38:in `block in run'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/cli/binstubs.rb:24:in `each'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/cli/binstubs.rb:24:in `run'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/cli.rb:321:in `binstubs'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/cli.rb:27:in `dispatch'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/cli.rb:18:in `start'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/exe/bundle:30:in `block in <top (required)>'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/site_ruby/2.4.0/bundler/friendly_errors.rb:122:in `with_friendly_errors'
  /Users/jared/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/exe/bundle:22:in `<top (required)>'
  /Users/jared/.rbenv/versions/2.4.2/bin/bundle:23:in `load'
  /Users/jared/.rbenv/versions/2.4.2/bin/bundle:23:in `<main>'
```

## Environment

```
Bundler       1.16.0
  Platforms   ruby, x86_64-darwin-16
Ruby          2.4.2p198 (2017-09-14 revision 59899) [x86_64-darwin16]
  Full Path   /Users/jared/.rbenv/versions/2.4.2/bin/ruby
  Config Dir  /Users/jared/.rbenv/versions/2.4.2/etc
RubyGems      2.7.0
  Gem Home    /Users/jared/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0
  Gem Path    /Users/jared/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0:/Users/jared/.gem/ruby/2.4.0
  User Path   /Users/jared/.gem/ruby/2.4.0
  Bin Dir     /Users/jared/.rbenv/versions/2.4.2/bin
OpenSSL       
  Compiled    OpenSSL 1.0.2l  25 May 2017
  Loaded      OpenSSL 1.0.2m  2 Nov 2017
  Cert File   /usr/local/etc/openssl/cert.pem
  Cert Dir    /usr/local/etc/openssl/certs
Tools         
  Git         2.15.0
  RVM         not installed
  rbenv       rbenv 1.1.1-6-g2d7cefe

  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2017-11-03
Git SHA           2d7cefe
Released Version  false
```

## Bundler settings

```
jobs
  Set for the current user (/Users/jared/.bundle/config): "4"
gem.coc
  Set for the current user (/Users/jared/.bundle/config): true
gem.mit
  Set for the current user (/Users/jared/.bundle/config): true
gem.test
  Set for the current user (/Users/jared/.bundle/config): "rspec"
ignore_messages
  Set for the current user (/Users/jared/.bundle/config): true
disable_version_check
  Set for the current user (/Users/jared/.bundle/config): false
gemfile
  Set via BUNDLE_GEMFILE: "/Users/jared/git/learn_webpacker/learn_webpacker/Gemfile"
bin_path
  Set via BUNDLE_BIN_PATH: "/Users/jared/.rbenv/versions/2.4.2/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/exe/bundle"
```

### Gemfile

```ruby
source 'https://rubygems.org'

git_source(:github) do |repo_name|
  repo_name = "#{repo_name}/#{repo_name}" unless repo_name.include?("/")
  "https://github.com/#{repo_name}.git"
end


# Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
gem 'rails', '~> 5.1.4'
# Use sqlite3 as the database for Active Record
gem 'sqlite3'
# Use Puma as the app server
gem 'puma', '~> 3.7'
# Use SCSS for stylesheets
gem 'sass-rails', '~> 5.0'
# Use Uglifier as compressor for JavaScript assets
gem 'uglifier', '>= 1.3.0'
# Transpile app-like JavaScript. Read more: https://github.com/rails/webpacker
gem 'webpacker'
# See https://github.com/rails/execjs#readme for more supported runtimes
# gem 'therubyracer', platforms: :ruby

# Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
gem 'jbuilder', '~> 2.5'
# Use ActiveModel has_secure_password
# gem 'bcrypt', '~> 3.1.7'

# Use Capistrano for deployment
# gem 'capistrano-rails', group: :development

group :development, :test do
  # Call 'byebug' anywhere in the code to stop execution and get a debugger console
  gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
  # Adds support for Capybara system testing and selenium driver
  gem 'capybara', '~> 2.13'
  gem 'selenium-webdriver'
end

group :development do
  # Access an IRB console on exception pages or by using <%= console %> anywhere in the code.
  gem 'web-console', '>= 3.3.0'
  gem 'listen', '>= 3.0.5', '< 3.2'
end

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    actioncable (5.1.4)
      actionpack (= 5.1.4)
      nio4r (~> 2.0)
      websocket-driver (~> 0.6.1)
    actionmailer (5.1.4)
      actionpack (= 5.1.4)
      actionview (= 5.1.4)
      activejob (= 5.1.4)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 2.0)
    actionpack (5.1.4)
      actionview (= 5.1.4)
      activesupport (= 5.1.4)
      rack (~> 2.0)
      rack-test (>= 0.6.3)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.1.4)
      activesupport (= 5.1.4)
      builder (~> 3.1)
      erubi (~> 1.4)
      rails-dom-testing (~> 2.0)
      rails-html-sanitizer (~> 1.0, >= 1.0.3)
    activejob (5.1.4)
      activesupport (= 5.1.4)
      globalid (>= 0.3.6)
    activemodel (5.1.4)
      activesupport (= 5.1.4)
    activerecord (5.1.4)
      activemodel (= 5.1.4)
      activesupport (= 5.1.4)
      arel (~> 8.0)
    activesupport (5.1.4)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (~> 0.7)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    addressable (2.5.2)
      public_suffix (>= 2.0.2, < 4.0)
    arel (8.0.0)
    bindex (0.5.0)
    builder (3.2.3)
    byebug (9.1.0)
    capybara (2.15.4)
      addressable
      mini_mime (>= 0.1.3)
      nokogiri (>= 1.3.3)
      rack (>= 1.0.0)
      rack-test (>= 0.5.4)
      xpath (~> 2.0)
    childprocess (0.8.0)
      ffi (~> 1.0, >= 1.0.11)
    concurrent-ruby (1.0.5)
    crass (1.0.2)
    erubi (1.7.0)
    execjs (2.7.0)
    ffi (1.9.18)
    globalid (0.4.1)
      activesupport (>= 4.2.0)
    i18n (0.9.1)
      concurrent-ruby (~> 1.0)
    jbuilder (2.7.0)
      activesupport (>= 4.2.0)
      multi_json (>= 1.2)
    listen (3.1.5)
      rb-fsevent (~> 0.9, >= 0.9.4)
      rb-inotify (~> 0.9, >= 0.9.7)
      ruby_dep (~> 1.2)
    loofah (2.1.1)
      crass (~> 1.0.2)
      nokogiri (>= 1.5.9)
    mail (2.7.0)
      mini_mime (>= 0.1.1)
    method_source (0.9.0)
    mini_mime (0.1.4)
    mini_portile2 (2.3.0)
    minitest (5.10.3)
    multi_json (1.12.2)
    nio4r (2.1.0)
    nokogiri (1.8.1)
      mini_portile2 (~> 2.3.0)
    public_suffix (3.0.0)
    puma (3.10.0)
    rack (2.0.3)
    rack-proxy (0.6.2)
      rack
    rack-test (0.7.0)
      rack (>= 1.0, < 3)
    rails (5.1.4)
      actioncable (= 5.1.4)
      actionmailer (= 5.1.4)
      actionpack (= 5.1.4)
      actionview (= 5.1.4)
      activejob (= 5.1.4)
      activemodel (= 5.1.4)
      activerecord (= 5.1.4)
      activesupport (= 5.1.4)
      bundler (>= 1.3.0)
      railties (= 5.1.4)
      sprockets-rails (>= 2.0.0)
    rails-dom-testing (2.0.3)
      activesupport (>= 4.2.0)
      nokogiri (>= 1.6)
    rails-html-sanitizer (1.0.3)
      loofah (~> 2.0)
    railties (5.1.4)
      actionpack (= 5.1.4)
      activesupport (= 5.1.4)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (12.2.1)
    rb-fsevent (0.10.2)
    rb-inotify (0.9.10)
      ffi (>= 0.5.0, < 2)
    ruby_dep (1.5.0)
    rubyzip (1.2.1)
    sass (3.5.3)
      sass-listen (~> 4.0.0)
    sass-listen (4.0.0)
      rb-fsevent (~> 0.9, >= 0.9.4)
      rb-inotify (~> 0.9, >= 0.9.7)
    sass-rails (5.0.6)
      railties (>= 4.0.0, < 6)
      sass (~> 3.1)
      sprockets (>= 2.8, < 4.0)
      sprockets-rails (>= 2.0, < 4.0)
      tilt (>= 1.1, < 3)
    selenium-webdriver (3.7.0)
      childprocess (~> 0.5)
      rubyzip (~> 1.0)
    sprockets (3.7.1)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.2.1)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
    sqlite3 (1.3.13)
    thor (0.20.0)
    thread_safe (0.3.6)
    tilt (2.0.8)
    tzinfo (1.2.4)
      thread_safe (~> 0.1)
    uglifier (3.2.0)
      execjs (>= 0.3.0, < 3)
    web-console (3.5.1)
      actionview (>= 5.0)
      activemodel (>= 5.0)
      bindex (>= 0.4.0)
      railties (>= 5.0)
    webpacker (3.0.2)
      activesupport (>= 4.2)
      rack-proxy (>= 0.6.1)
      railties (>= 4.2)
    websocket-driver (0.6.5)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.2)
    xpath (2.1.0)
      nokogiri (~> 1.3)

PLATFORMS
  ruby

DEPENDENCIES
  byebug
  capybara (~> 2.13)
  jbuilder (~> 2.5)
  listen (>= 3.0.5, < 3.2)
  puma (~> 3.7)
  rails (~> 5.1.4)
  sass-rails (~> 5.0)
  selenium-webdriver
  sqlite3
  tzinfo-data
  uglifier (>= 1.3.0)
  web-console (>= 3.3.0)
  webpacker

BUNDLED WITH
   1.16.0
```

--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

**Comment by colby-swandale** [comment]
I can't recreate this issue, can you try `gem pristine bundler -v '1.16.0'` and try again.

**Comment by ju5t** [comment]
I'm not that into Ruby and I only use it because it's what Puppet uses, but I've been able to reproduce this on a clean Docker image. I don't know exactly how `rubygems-update` and `bundler` interface with each other but it only happens when I upgrade `rubygems-update` to the latest version.

```
docker run --rm -it ruby:2.3 bash
```

Before:

```
# gem list

*** LOCAL GEMS ***

bigdecimal (default: 1.2.8)
bundler (1.16.0)
did_you_mean (1.0.0)
io-console (default: 0.4.5)
json (default: 1.8.3.1)
minitest (5.8.5)
net-telnet (0.1.1)
power_assert (0.2.6)
psych (default: 2.1.0.1)
rake (10.4.2)
rdoc (default: 4.2.1)
rubygems-update (2.6.14)
test-unit (3.1.5)
```

After `gem update --system` I get:

```
# gem list

*** LOCAL GEMS ***

bigdecimal (default: 1.2.8)
bundler (default: 1.16.0)
did_you_mean (1.0.0)
io-console (default: 0.4.5)
json (default: 1.8.3.1)
minitest (5.8.5)
net-telnet (0.1.1)
power_assert (0.2.6)
psych (default: 2.1.0.1)
rake (10.4.2)
rdoc (default: 4.2.1)
rubygems-update (2.7.1, 2.6.14)
test-unit (3.1.5)
```

Once I've upgraded I run into similar problems with the following Gemfile.

```
cat <<GEMFILE > Gemfile
source ENV['GEM_SOURCE'] || 'https://rubygems.org'

puppetversion = ENV.key?('PUPPET_VERSION') ? ENV['PUPPET_VERSION'] : ['>= 3.3']
gem 'facter', '<= 2.4.6', '>= 1.7.0'
gem 'puppet', puppetversion
gem 'puppet-lint', '>= 1.0.0'
gem 'puppetlabs_spec_helper', '>= 1.0.0'
gem 'rspec-puppet', '<= 2.5.0'
gem 'rspec-puppet-facts', '< 1.8.0'
gem 'safe_yaml', '~> 1.0.4'
gem 'simplecov', require: false
gem 'simplecov-console', require: false

# handle old ruby versions
if RUBY_VERSION < '2.0.0'
  gem 'metadata-json-lint', '1.1.0'
  gem 'rubocop', '<= 0.41.2'
else
  gem 'metadata-json-lint', '< 2.0.0'
  gem 'rubocop', '>= 0.48.1'
end

group :system_tests do
  gem 'beaker', '<= 2.51.0' if RUBY_VERSION < '2.2.5'
  gem 'beaker-puppet_install_helper'
  gem 'beaker-rspec'
  gem 'nokogiri', '< 1.7.0' if RUBY_VERSION < '2.1.0'
  gem 'public_suffix', '<= 1.4.6' if RUBY_VERSION < '2.0.0'
end
GEMFILE
```

As `bundle install --path vendor/bundle` results in:

```
Errno::ENOENT: No such file or directory @ rb_sysopen - /usr/local/lib/ruby/site_ruby/2.3.0/bundler/templates/Executable
An error occurred while installing rake (12.2.1), and Bundler cannot continue.
Make sure that `gem install rake -v '12.2.1'` succeeds before bundling.
```

If I leave `rubygems-update` at `2.6.14` everything works as expected.

**Comment by colby-swandale** [comment]
Thanks @ju5t, i can reproduce this now, this issue is tied to rubygems 2.7.

\cc @segiddins

**Comment by coderanger** [comment]
Getting this error across most of my Travis builds (though oddly not all), https://travis-ci.org/poise/poise/jobs/297201059

**Comment by ivan-kolmychek** [comment]
Getting this error on a lot of our CI builds as well. Can reproduce it locally too by updating to the latest `ruby:2.4`.

**Comment by nullnull** [comment]
I think I have same problem. Minimum example with docker to reproduce this bug is following.

```sh
$ cat Dockerfile
FROM ruby:2.4.1
RUN gem update --system
RUN echo "source 'https://rubygems.org'" > Gemfile
RUN echo "gem 'rails', '5.1.0'" >> Gemfile
RUN bundle install
$ docker build .
```

**Comment by mgathu1** [comment]
@ju5t solution helped me. I downgraded to rubygems 2.6.14 and all was smooth sailing with my docker build.

**Comment by colby-swandale** [comment]
This should be fixed by https://github.com/rubygems/rubygems/pull/2065

**Comment by segiddins** [comment]
Closing as fixed by RubyGems 2.7.2

**Comment by jaredbeck** [comment]
Hi, OP here, just confirming that the combination of `gem 2.7.2` and `bundler 1.16.0` is capable of installing rails. Thanks, everyone who worked on this.
